core/src/unprocessed_packet_batches.rs

```diff
@@ -114,211 +147,273 @@ impl Ord for DeserializedPacket {
+        }
+        // purge empty batches
+        self.purge_empty_batch();
     }
 
-    pub fn retain<F>(&mut self, mut f: F)
-    where
-        F: FnMut(&mut DeserializedPacket) -> bool,
-    {
-        // TODO: optimize this only when number of packets
-        // with oudated blockhash is high
+    fn clear(&mut self) {
         self.packet_priority_queue.clear();
-        self.message_hash_to_transaction
-            .retain(|_k, deserialized_packet| {
-                let should_retain = f(deserialized_packet);
-                if should_retain {
-                    let priority_queue_entry =
-                        PacketPriorityQueueEntry::from_packet(deserialized_packet);
-                    self.packet_priority_queue.push(priority_queue_entry);
-                }
-                should_retain
-            })
+        self.packet_batches.clear();
     }
 
-    pub fn len(&self) -> usize {
+    // returns number of unprocessed packets in buffer
+    fn len(&self) -> usize {
         self.packet_priority_queue.len()
     }
 
-    pub fn is_empty(&self) -> bool {
+    fn is_empty(&self) -> bool {
         self.packet_priority_queue.is_empty()
     }
-
-    fn push_internal(&mut self, deserialized_packet: DeserializedPacket) {
-        let priority_queue_entry = PacketPriorityQueueEntry::from_packet(&deserialized_packet);
-
-        // Push into the priority queue
-        self.packet_priority_queue.push(priority_queue_entry);
-
-        // Keep track of the original packet in the tracking hashmap
-        self.message_hash_to_transaction
-            .insert(deserialized_packet.message_hash(), deserialized_packet);
+    fn capacity(&self) -> usize {
+        self.packet_priority_queue.capacity()
     }
+}
 
-    /// Returns the popped minimum packet from the priority queue.
-    fn push_pop_min(&mut self, deserialized_packet: DeserializedPacket) -> DeserializedPacket {
-        let priority_queue_entry = PacketPriorityQueueEntry::from_packet(&deserialized_packet);
-
-        // Push into the priority queue
-        let popped_priority_queue_entry = self
-            .packet_priority_queue
-            .push_pop_min(priority_queue_entry);
+impl UnprocessedPacketBatches {
+    pub fn with_capacity(capacity: usize) -> Self {
+        UnprocessedPacketBatches {
+            packet_priority_queue: MinMaxHeap::with_capacity(capacity),
+            packet_batches: VecDeque::with_capacity(capacity),
+            batch_limit: capacity,
+        }
+    }
 
-        // Keep track of the original packet in the tracking hashmap
-        self.message_hash_to_transaction
-            .insert(deserialized_packet.message_hash(), deserialized_packet);
+    pub fn set_all_packets_forwarded(&mut self) {
+        for batch in self.packet_batches.iter() {
+            for (_k, deserialized_packet) in batch.borrow_mut().deserialized_packets.iter_mut() {
+                if let Some(packet) = Rc::get_mut(deserialized_packet) {
```

Comment:
this `Rc::get_mut` prolly not the ideal way to modify a ref-counted object. @carllin mentioned early to properly does, it'd be `Rc<RefCell<>>`.